### PR TITLE
Support CREATE / DROP database commands from any node

### PR DESCRIPTION
**Pull request by onurctirtir**
DESCRIPTION: Adds support for issuing `CREATE`/`DROP` DATABASE commands from worker nodes

With this commit, we allow issuing CREATE / DROP DATABASE commands from worker nodes too.
As in #7278, this is not allowed when the coordinator is not added to metadata because
we don't ever sync metadata changes to coordinator when adding coordinator to
the metadata via `SELECT citus_set_coordinator_host('<hostname>')`, or equivalently, via
`SELECT citus_add_node(<coordinator_node_name>, <coordinator_node_port>, 0)`.

We serialize database management commands by acquiring a Citus specific advisory lock
on the first primary worker node if there are any workers in the cluster. As opposed to
what we've done in https://github.com/citusdata/citus/pull/7278 for role management
commands, we try to avoid from running into distributed deadlocks as much as possible.
This is because, while distributed deadlocks that can happen around role management
commands can be detected by Citus, this is not the case for database management commands
because most of them cannot be run inside in a transaction block. In that case, Citus
cannot even detect the distributed deadlock because the command is not part of a
distributed transaction at all, then the command execution might not return the control
back to the user for an indefinite amount of time.

### CODE REVIEW

File: src/test/regress/sql/columnar_drop.sql [hunk 1]
```diff
@@ -15,23 +15,27 @@
...
+DROP TABLE IF EXISTS contestant;
+DROP TABLE IF EXISTS contestant_compressed;
+RESET client_min_messages;
 
 -- make sure DROP deletes metadata
-SELECT :columnar_stripes_before_drop - count(distinct storage_id) FROM columnar.stripe;
+WITH cte AS (SELECT distinct relation as distinct_rel FROM columnar.stripe)
+SELECT
+  (SELECT COUNT(*) > 0 FROM cte WHERE distinct_rel::regclass::text like '%contestant%') AS contestant_rel_exists,
+  (SELECT COUNT(*) > 0 FROM cte WHERE distinct_rel::regclass::text like '%contestant_compressed%') AS contestant_compressed_rel_exists;
```
Comment (onurctirtir) [inline src/test/regress/sql/columnar_drop.sql:28]: This PR needs to update output of this test hence this brings the necessity of making this test idempotent. Those changes are done to make this test idempotent.

Comment (onurctirtir) [reply src/test/regress/sql/columnar_drop.sql:28]: That's also why this PR updates run_test.py.

File: src/test/regress/sql/multi_utility_warnings.sql [hunk 2]
```diff
@@ -6,3 +6,4 @@
 -- databases.
 
 CREATE DATABASE new_database;
+DROP DATABASE new_database;
```
Comment (onurctirtir) [inline src/test/regress/sql/multi_utility_warnings.sql:9]: same here, this is needed to make this test idempotent because this PR needs to update its output for the sake of the changes introduced here

**Review by onurctirtir — COMMENTED** [review]
Hmm, we have a strange situation where we can't even detect the deadlock?

```sql
select citus_add_node('localhost', 9700, 0);
select run_command_on_all_nodes('alter system set citus.enable_create_database_propagation to on');
select run_command_on_all_nodes('select pg_reload_conf()');
select run_command_on_all_nodes('create database onur');
```

Or equivalently, can use the following script too in order to get rid of the call made to `run_command_on_all_nodes()`:

```sh
#!/bin/bash

# make sure that coordinator is added to metadata and the GUC is enabled
psql -p 9700 -c "SELECT citus_add_node('localhost', 9700, 0)"
psql -p 9700 -c "SELECT run_command_on_all_nodes('alter system set citus.enable_create_database_propagation to on')"
psql -p 9700 -c "SELECT run_command_on_all_nodes('select pg_reload_conf()')"

dbname="$1"

pids=()

for i in `seq 0 2`; do
    psql -p 970${i} -c "create database $dbname" &
    pids+=($!)
done

# wait for all pids
for pid in ${pids[*]}; do
    wait $pid
done
```

File: src/backend/distributed/commands/database.c [hunk 3]
```diff
@@ -302,7 +302,7 @@ PreprocessCreateDatabaseStmt(Node *node, const char *queryString,
 		return NIL;
 	}
 
-	EnsureCoordinator();
+	EnsurePropagationToCoordinator();
```
Comment (gurkanindibay) [inline src/backend/distributed/commands/database.c:305]: Are you sure that EnsurePropagationToCoordinator covered?

Comment (onurctirtir) [reply src/backend/distributed/commands/database.c:305]: Indeed I can see the error thrown in the test file but maybe codecov is complaining about some branch coverage problems (as we've discussed yesterday) that I'm not aware of.

Comment (onurctirtir) [reply src/backend/distributed/commands/database.c:305]: As we can see in `create_drop_database_propagation.out`:

```sql
\c - - - :worker_1_port
SET citus.enable_create_database_propagation TO ON;
-- fails because coordinator is not added into metadata
DROP DATABASE another_db_created_from_worker;
ERROR:  coordinator is not added to the metadata
HINT:  Use SELECT citus_set_coordinator_host('<hostname>') to configure the coordinator hostname
```

File: src/test/regress/expected/create_drop_database_propagation.out [hunk 4]
```diff
@@ -941,6 +937,164 @@ SELECT * FROM public.check_database_on_all_nodes('test_db') ORDER BY node_type;
...
+ worker node (local)  | {"database_properties": {"datacl": null, "datname": "db_created_from_worker", "datctype": "C", "encoding": "UTF8", "datcollate": "C", "tablespace": "ts-needs\\!escape", "daticurules": null, "datallowconn": false, "datconnlimit": 42, "daticulocale": null, "datistemplate": false, "database_owner": "create_drop_db_test_user", "datcollversion": null, "datlocprovider": "c"}, "pg_dist_object_record_for_db_exists": true, "stale_pg_dist_object_record_for_a_db_exists": false}
+ worker node (remote) | {"database_properties": {"datacl": null, "datname": "db_created_from_worker", "datctype": "C", "encoding": "UTF8", "datcollate": "C", "tablespace": "ts-needs\\!escape", "daticurules": null, "datallowconn": false, "datconnlimit": 42, "daticulocale": null, "datistemplate": false, "database_owner": "create_drop_db_test_user", "datcollversion": null, "datlocprovider": "c"}, "pg_dist_object_record_for_db_exists": true, "stale_pg_dist_object_record_for_a_db_exists": false}
+(3 rows)
+
+DROP DATABASE db_created_from_worker;
```
Comment (gurkanindibay) [inline src/test/regress/expected/create_drop_database_propagation.out:990]: I couldn't understand the aim of this test. There is only one local db already. So, having GUC on with just single DB proves nothing IMO

File: src/test/regress/expected/create_drop_database_propagation.out [hunk 5]
```diff
@@ -941,6 +937,164 @@ SELECT * FROM public.check_database_on_all_nodes('test_db') ORDER BY node_type;
...
+SELECT * FROM public.check_database_on_all_nodes('local_worker_db') ORDER BY node_type;
+      node_type       |                                                              result
+---------------------------------------------------------------------
+ coordinator (remote) | {"database_properties": null, "pg_dist_object_record_for_db_exists": false, "stale_pg_dist_object_record_for_a_db_exists": false}
+ worker node (local)  | {"database_properties": null, "pg_dist_object_record_for_db_exists": false, "stale_pg_dist_object_record_for_a_db_exists": false}
+ worker node (remote) | {"database_properties": null, "pg_dist_object_record_for_db_exists": false, "stale_pg_dist_object_record_for_a_db_exists": false}
+(3 rows)
+
+SET citus.enable_create_database_propagation TO OFF;
+CREATE DATABASE local_worker_db;
```
Comment (gurkanindibay) [inline src/test/regress/expected/create_drop_database_propagation.out:1010]: I couldn't understand the aim of this test. There is only one local db already. So, having GUC on with just single DB proves nothing IMO

File: src/test/regress/expected/create_drop_database_propagation.out (hunk 4 above)
Comment (gurkanindibay) [inline src/test/regress/expected/create_drop_database_propagation.out:990]: I couldn't understand the aim of this test. There is only one local db already. So, having GUC on with just single DB proves nothing IMO

File: src/test/regress/expected/create_drop_database_propagation.out (hunk 5 above)
Comment (gurkanindibay) [inline src/test/regress/expected/create_drop_database_propagation.out:1010]: I couldn't understand the aim of this test. There is only one local db already. So, having GUC off with just single DB proves nothing IMO

**Comment by gurkanindibay** [comment]
A good description is needed. AFAIU, Removing coordinator limits the propagation. It should be clearly explained in PR description.

File: src/test/regress/expected/create_drop_database_propagation.out (hunk 5 above)
Comment (onurctirtir) [reply src/test/regress/expected/create_drop_database_propagation.out:1010]: wanted to show what happens when the command is issued from a worker while the GUC is off, i.e., whether we preserve the legacy behaivour

**Comment by onurctirtir** [comment]
> Hmm, we have a strange situation where we can't even detect the deadlock?
> 
> ```sql
> select citus_add_node('localhost', 9700, 0);
> select run_command_on_all_nodes('alter system set citus.enable_create_database_propagation to on');
> select run_command_on_all_nodes('select pg_reload_conf()');
> select run_command_on_all_nodes('create database onur');
> ```
> 
> Or equivalently, can use the following script too in order to get rid of the call made to `run_command_on_all_nodes()`:
> 
> ```shell
> psql -p 9700 -c "SELECT citus_add_node('localhost', 9700, 0)"
> psql -p 9700 -c "SELECT run_command_on_all_nodes('alter system set citus.enable_create_database_propagation to on')"
> psql -p 9700 -c "SELECT run_command_on_all_nodes('select pg_reload_conf()')"
> 
> dbname="$1"
> 
> pids=()
> 
> for i in `seq 0 2`; do
>     psql -p 970${i} -c "create database $dbname" &
>     pids+=($!)
> done
> 
> for pid in ${pids[*]}; do
>     wait $pid
> done
> ```

First of all, deadlock detection code doesn't care about the queries
that are not part of a distributed transaction. See the occurrences of
`bool onlyDistributedTx = true` in `distributed_deadlock_detection.c`

Setting `onlyDistributedTx` to `true` doesn't solve this problem because
`AddWaitEdge()` then cannot track initiator node for a backend that is not
part of a distributed transaction. This is because, we call
`assign_distributed_transaction_id()` only for the backends that are part
of a distributed transaction and `AddWaitEdge()` deduces the initiator
node thanks what `assign_distributed_transaction_id()` saves in backend data.

And when this info is not provided, distributed deadlock detector cannot
even associate a remote backend to its originating node and cannot detect
the cycles, hence cannot detect the distributed deadlock.

File: src/backend/distributed/serialize_distributed_ddls.c [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+
+/*
+ * SerializeDistributedDDLsOnObjectClass serializes distributed DDLs that
+ * target given object class by acquiring a Citus specific advisory lock
+ * on the first primary worker node.
+ *
+ * The lock is acquired via a coordinated transaction. For this reason,
+ * it automatically gets released when (maybe implicit) transaction on
+ * current server commits or rolls back.
```
Comment (onurctirtir) [inline src/backend/distributed/serialize_distributed_ddls.c:31]: Verify if that's really the case for create database.

Do we by mistake commit right after SET LOCAL thing?

File: src/backend/distributed/commands/database.c [hunk 7]
```diff
@@ -328,7 +330,7 @@ PostprocessCreateDatabaseStmt(Node *node, const char *queryString)
...
```
Comment (onurctirtir) [inline src/backend/distributed/commands/database.c:333]: missing SerializeDistributedDDLsOnObjectClass(OCLASS_DATABASE);

File: src/backend/distributed/commands/database.c [hunk 8]
```diff
@@ -366,7 +368,7 @@ PreprocessDropDatabaseStmt(Node *node, const char *queryString,
...
```
Comment (onurctirtir) [inline src/backend/distributed/commands/database.c:371]: missing SerializeDistributedDDLsOnObjectClass(OCLASS_DATABASE);

File: src/backend/distributed/commands/database.c [hunk 9]
```diff
@@ -303,7 +304,8 @@ PreprocessCreateDatabaseStmt(Node *node, const char *queryString,
...
+	SerializeDistributedDDLsOnObjectClass(OCLASS_DATABASE);
```
Comment (onurctirtir) [inline src/backend/distributed/commands/database.c:308]: move this to a place right before return?

File: src/backend/distributed/metadata/metadata_sync.c [hunk 10]
```diff
@@ -4049,6 +4050,29 @@ citus_internal_database_command(PG_FUNCTION_ARGS)
...
+{
+	CheckCitusVersion(ERROR);
+
+	PG_ENSURE_ARGNOTNULL(0, "object_class");
+	ObjectClass objectClass = PG_GETARG_INT32(0);
+
+	if (!ShouldSkipMetadataChecks())
+	{
+		EnsureCitusInitiatedOperation();
+	}
```
Comment (onurctirtir) [inline src/backend/distributed/metadata/metadata_sync.c:4068]: needs tests

File: src/backend/distributed/sql/udfs/citus_internal_acquire_citus_advisory_object_class_lock/12.2-1.sql [hunk 11]
```diff
@@ -0,0 +1,5 @@
+CREATE OR REPLACE FUNCTION pg_catalog.citus_internal_acquire_citus_advisory_object_class_lock(objectClass int)
```
Comment (onurctirtir) [inline src/backend/distributed/sql/udfs/citus_internal_acquire_citus_advisory_object_class_lock/12.2-1.sql:1]: missing comment & `Comment ON` clause?

File: src/include/distributed/resource_lock.h [hunk 12]
```diff
@@ -142,6 +144,48 @@ typedef enum CitusOperations
...
+		default:
+			ereport(ERROR, (errmsg("unexpected object class: %d", objectClass)));
+	}
+}
+
+
+/*
+ * Reuse advisory lock, but with different, unused field 4 (14).
+ *
+ * Automatically sets databaseId to MyDatabaseId if the object class is
```
Comment (onurctirtir) [inline src/include/distributed/resource_lock.h:173]:
```suggestion
 * Automatically sets databaseId to InvalidOid if the object class is
```

File: src/include/distributed/resource_lock.h [hunk 13]
```diff
@@ -142,6 +144,48 @@ typedef enum CitusOperations
...
+ * node-wide, i.e., that is not bound to a particular database but to
+ * whole server. If the object class is not node-wide, sets databaseId
+ * to InvalidOid.
```
Comment (onurctirtir) [inline src/include/distributed/resource_lock.h:176]:
```suggestion
 * to MyDatabaseId.
```

**Review by onurctirtir — COMMENTED** [review]
- [ ] Improve PR description on SerializeDistributedDDLsOnObjectClass.
- [ ] Improve PR description on https://github.com/citusdata/citus/pull/7359#pullrequestreview-1742403646.
- [ ] Add more tests related to concurrency.

File: src/backend/distributed/serialize_distributed_ddls.c (hunk 6 above)
Comment (onurctirtir) [reply src/backend/distributed/serialize_distributed_ddls.c:31]: no

File: src/backend/distributed/metadata/metadata_sync.c [hunk 14]
```diff
@@ -4279,6 +4303,22 @@ DeletePlacementMetadataCommand(uint64 placementId)
...
+ * AcquireCitusAdvisoryObjectClassLockCommand returns a command to call
+ * citus_internal_acquire_citus_advisory_object_class_lock().
+ */
+char *
+AcquireCitusAdvisoryObjectClassLockCommand(ObjectClass objectClass)
+{
+	StringInfo command = makeStringInfo();
+	appendStringInfo(command,
+					 "SELECT pg_catalog.citus_internal_acquire_citus_advisory_object_class_lock(%d)",
+					 objectClass);
```
Comment (thanodnl) [inline src/backend/distributed/metadata/metadata_sync.c:4316]: I wonder if this is more readable when we cast the objectClass from the enum to an int explicitly. It reads weird to format a custom type as %d. Alternatively we could add a macro somewhere like `OBJECTCLASS_FORMAT` in the same manner that other format symbols are aliassed.

**Review by thanodnl — COMMENTED** [review]
Looks good overal

File: src/backend/distributed/commands/database.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (onurctirtir) [reply src/backend/distributed/commands/database.c:341]: I think this is fine as long as we only call this function for DDLs, I'd prefer keeping this for postprocess too for safety.

File: src/include/distributed/resource_lock.h [hunk 15]
```diff
@@ -142,6 +144,48 @@ typedef enum CitusOperations
...
+{
+	switch (objectClass)
+	{
+		case OCLASS_DATABASE:
+		{
+			return true;
+		}
+
+		default:
+			ereport(ERROR, (errmsg("unexpected object class: %d", objectClass)));
```
[the remark this answers is not in the document]
Comment (onurctirtir) [reply src/include/distributed/resource_lock.h:165]: I was imagining that this function -in future- would return return false / true depending on -known- input and throw an error otherwise for safety. And given that today the only expect object class to be OCLASS_DATABASE, ended up such a code that today might not make much sense.

Do you think should we instead return false for now?

[the remark this answers is not in the document]
Comment (onurctirtir) [reply src/include/distributed/resource_lock.h:165]: just seen https://github.com/citusdata/citus/pull/7359#discussion_r1410803546, alright.

File: src/backend/distributed/sql/udfs/citus_internal_acquire_citus_advisory_object_class_lock/latest.sql [hunk 16]
```diff
@@ -0,0 +1,5 @@
+CREATE OR REPLACE FUNCTION pg_catalog.citus_internal_acquire_citus_advisory_object_class_lock(objectClass int)
```
Comment (JelteF) [inline src/backend/distributed/sql/udfs/citus_internal_acquire_citus_advisory_object_class_lock/latest.sql:1]: Should use the citus_internal schema
```suggestion
CREATE OR REPLACE FUNCTION citus_internal.acquire_citus_advisory_object_class_lock(objectClass int)
```

File: src/include/distributed/resource_lock.h [hunk 17]
```diff
@@ -142,6 +144,75 @@ typedef enum CitusOperations
...
+	 * this assertion check based on latest supported major Postgres version.
+	 */
+	StaticAssertStmt(PG_MAJORVERSION_NUM <= 16,
+					 "better to check if any of newly added ObjectClass'es are node-wide");
+
+	switch (objectClass)
+	{
+		case OCLASS_ROLE:
+		case OCLASS_DATABASE:
+		case OCLASS_SUBSCRIPTION:
```
Comment (JelteF) [inline src/include/distributed/resource_lock.h:179]: Is subscription really node wide? Afaik they are per database.

File: src/test/regress/expected/create_drop_database_propagation.out [hunk 18]
```diff
@@ -932,6 +928,259 @@ SELECT * FROM public.check_database_on_all_nodes('test_db') ORDER BY node_type;
...
+ ?column?
+---------------------------------------------------------------------
+        1
+(1 row)
+
+\c - - - :worker_1_port
+CREATE DATABASE local_worker_db;
+NOTICE:  Citus partially supports CREATE DATABASE for distributed databases
+DETAIL:  Citus does not propagate CREATE DATABASE command to other nodes
+HINT:  You can manually create a database and its extensions on other nodes.
```
Comment (JelteF) [inline src/test/regress/expected/create_drop_database_propagation.out:943]: Sidenote: Should we update this hint message to include a hint about citus.enable_create_database_propagation? In any case this would be better to do in a separate PR.

File: src/test/regress/expected/create_drop_database_propagation.out [hunk 19]
```diff
@@ -932,6 +928,259 @@ SELECT * FROM public.check_database_on_all_nodes('test_db') ORDER BY node_type;
...
+(3 rows)
+
+SELECT pg_sleep(0.1);
+ pg_sleep
+---------------------------------------------------------------------
+
+(1 row)
+
+-- only one of them succeeds and we don't run into a distributed deadlock
+SELECT COUNT(*)=1 FROM run_command_on_all_nodes('CREATE DATABASE concurrent_create_db') WHERE success;
```
Comment (JelteF) [inline src/test/regress/expected/create_drop_database_propagation.out:1131]: For debugging purposes it's nicer to show the actual value instead of true/false
```suggestion
SELECT COUNT(*) FROM run_command_on_all_nodes('CREATE DATABASE concurrent_create_db') WHERE success;
```

File: src/test/regress/expected/create_drop_database_propagation.out [hunk 20]
```diff
@@ -932,6 +928,259 @@ SELECT * FROM public.check_database_on_all_nodes('test_db') ORDER BY node_type;
...
+        1
+(3 rows)
+
+SELECT pg_sleep(0.1);
+ pg_sleep
+---------------------------------------------------------------------
+
+(1 row)
+
+-- only one of them succeeds and we don't run into a distributed deadlock
```
Comment (JelteF) [inline src/test/regress/expected/create_drop_database_propagation.out:1130]: I think it would be good to have an isolation test showing that CREATE DATABASE commands for different databases block each other too.

File: src/include/distributed/resource_lock.h (hunk 17 above)
Comment (onurctirtir) [reply src/include/distributed/resource_lock.h:179]: I thought that they're node-wide after seeing this function, what do you think?

```c
bool
IsSharedRelation(Oid relationId)
{
	/* These are the shared catalogs (look for BKI_SHARED_RELATION) */
	if (relationId == AuthIdRelationId ||
		relationId == AuthMemRelationId ||
		relationId == DatabaseRelationId ||
		relationId == DbRoleSettingRelationId ||
		relationId == ParameterAclRelationId ||
		relationId == ReplicationOriginRelationId ||
		relationId == SharedDependRelationId ||
		relationId == SharedDescriptionRelationId ||
		relationId == SharedSecLabelRelationId ||
		relationId == SubscriptionRelationId ||
		relationId == TableSpaceRelationId)
		return true;
```

File: src/test/regress/expected/create_drop_database_propagation.out (hunk 18 above)
Comment (onurctirtir) [reply src/test/regress/expected/create_drop_database_propagation.out:943]: makes sense, let me update this in a separate PR

File: src/test/regress/expected/columnar_drop.out [hunk 21]
```diff
@@ -12,35 +12,39 @@
 -- 'postgres' directory is excluded from comparison to have the same result.
 -- store postgres database oid
 SELECT oid postgres_oid FROM pg_database WHERE datname = 'postgres' \gset
-SELECT count(distinct storage_id) AS columnar_stripes_before_drop FROM columnar.stripe \gset
 -- DROP columnar tables
-DROP TABLE contestant;
-DROP TABLE contestant_compressed;
+SET client_min_messages TO WARNING;
+DROP TABLE IF EXISTS contestant;
+DROP TABLE IF EXISTS contestant_compressed;
```
Comment (JelteF) [inline src/test/regress/expected/columnar_drop.out:18]: I feel like this is making the test quite useless on the second run, because it's not actually dropping anything anymore. I feel like it would be better to:
1. create new tables 
2. check they have stripes
3. drop them
4. check that there's no stripes anymore

Or another option would be to simply mark this test as non-repeatable.

File: src/backend/distributed/serialize_distributed_ddls.c [hunk 22]
```diff
@@ -0,0 +1,122 @@
...
+ * to call AcquireCitusAdvisoryObjectClassLock().
+ */
+Datum
+citus_internal_acquire_citus_advisory_object_class_lock(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+
+	PG_ENSURE_ARGNOTNULL(0, "object_class");
+	ObjectClass objectClass = PG_GETARG_INT32(0);
+
```
Comment (JelteF) [inline src/backend/distributed/serialize_distributed_ddls.c:42]: Let's put some security mechanism in place to check that not just any user can block create database DDL. I think the best would be to check if the user has CREATEDB privileges. If they do, then they are allowed to acquire the lock.

Comment (onurctirtir) [reply src/backend/distributed/serialize_distributed_ddls.c:42]: Thinking loudly ..

We're acquiring this lock not only for create database but also for alter / drop database commands to prevent the deadlocks that could happen, say, when issuing "create database X" & "alter database rename Y to X" concurrently.

For create database, it makes sense to ensure CREATEDB privilage. However for ALTER / DROP DATABASE, it would probably make more sense to check db ownership.

However, I'm not sure how we can do that in this UDF as it doesn't know about what type of command we're executing.

**Comment by onurctirtir** [comment]
TODO:
- [ ] Add isolation tests:
      https://github.com/citusdata/citus/pull/7359#discussion_r1437566199 
- [ ] Add tests around new permission checks.

File: src/backend/distributed/serialize_distributed_ddls.c [hunk 23]
```diff
@@ -0,0 +1,267 @@
...
+	char *qualifiedObjectName = PG_ARGISNULL(1) ? NULL : PG_GETARG_CSTRING(1);
```
Comment (onurctirtir) [inline src/backend/distributed/serialize_distributed_ddls.c:61]: Now this is based on "objectName" rather than "oid" because the previous approach was introducing a race condition where the subquery that we send to workers (to resolve the oid from the database) might return NULL oid in case another session concurrently dropped / renamed the object. And this was resulting in acquiring an incorrect lock (InvalidOid vs a real oid) even if it doesn't seem much harmful.


See https://github.com/citusdata/citus/actions/runs/7388502257/attempts/1#summary-20099571404.

File: src/backend/distributed/commands/database.c
Comment (onurctirtir) [inline src/backend/distributed/commands/database.c:1]: Not calling SerializeDistributedDDLsOnObjectClass for GRANT ON DATABASE .. TO ... commands for two reasons:
* It would mean mistakenly enforcing ownership for the database, though this is normally not the case in Postgres.
  See https://github.com/citusdata/citus/actions/runs/7396645191/attempts/1#summary-20122278383.
* We could add flag to the underlying UDF to workaround first problem but seems that GRANT commands are not likely to cause deadlock with other database commands.

**Comment by onurctirtir** [comment]
Failures in foreign_key_to_reference_shard_rebalance are not related to this PR and will be fixed by https://github.com/citusdata/citus/pull/7400.

File: src/backend/distributed/serialize_distributed_ddls.c [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+#include "distributed/deparse_shard_query.h"
+#include "distributed/metadata_sync.h"
+#include "distributed/resource_lock.h"
+#include "distributed/serialize_distributed_ddls.h"
+
+
+/*
+ * SerializeDistributedDDLsOnObjectClass serializes distributed DDLs that
+ * target given object class by acquiring a Citus specific advisory lock
+ * on the first primary worker node.
```
[the remark this answers is not in the document]
Comment (onurctirtir) [reply src/backend/distributed/serialize_distributed_ddls.c]: Discussed this in Teams and decided to move forward with using first primary worker node.

File: src/backend/distributed/sql/udfs/citus_internal_acquire_citus_advisory_object_class_lock/latest.sql (hunk 16 above)
Comment (onurctirtir) [reply src/backend/distributed/sql/udfs/citus_internal_acquire_citus_advisory_object_class_lock/latest.sql]: Moved this back to pg_catalog to avoid from reporting permission issues to regular users.

File: src/backend/distributed/utils/listutils.c [hunk 25]
```diff
@@ -289,3 +289,16 @@ list_filter_oid(List *list, bool (*keepElement)(Oid element))
 
 	return result;
 }
+
+
+/*
+ * CompareStringNodes returns the result of strcmp on given two String nodes.
+ */
+int
+CompareStringNodes(const void *leftElement, const void *rightElement)
```
Comment (JelteF) [inline src/backend/distributed/utils/listutils.c:298]: Seems to be dead code?

File: src/include/distributed/resource_lock.h (hunk 17 above)
Comment (JelteF) [reply src/include/distributed/resource_lock.h:179]: Okay, I checked. And it turns out that subscriptions are local to a database, but for some reason their table is not. Instead they use a global catalog table, where each entry has a `subdbid` column. I'm not entirely sure why this global-but-not-really-global design was chosen, but I think the reason might related to how global subscription views work, such as pg_stat_subscription. 

For our usecase we should not consider subscriptions a global object though. You cannot alter/drop a subscription that is in a different database. And creating subscriptions with the same name does not conflict across databases.

File: src/include/distributed/resource_lock.h [hunk 26]
```diff
@@ -142,6 +144,73 @@ typedef enum CitusOperations
...
+		case OCLASS_TBLSPACE:
+#if PG_VERSION_NUM >= PG_VERSION_15
+		case OCLASS_PARAMETER_ACL:
```
Comment (JelteF) [inline src/include/distributed/resource_lock.h:182]: hmm, this made me look what this was. Turns out we need to propagate some more global DDL that we don't support yet... GRANT/REVOKE ON PARAMETER

File: src/test/regress/expected/create_drop_database_propagation.out [hunk 27]
```diff
@@ -932,15 +928,350 @@ DROP ROLE propagated_role, non_propagated_role;
...
+CONTEXT:  while executing command on localhost:xxxxx
+ALTER DATABASE distributed_db SET timezone TO 'UTC';
+ERROR:  must be owner of database distributed_db
+CONTEXT:  while executing command on localhost:xxxxx
+ALTER DATABASE distributed_db RESET timezone;
+ERROR:  must be owner of database distributed_db
+CONTEXT:  while executing command on localhost:xxxxx
+GRANT ALL ON DATABASE distributed_db TO postgres;
+WARNING:  no privileges were granted for "distributed_db"
+RESET ROLE;
```
Comment (JelteF) [inline src/test/regress/expected/create_drop_database_propagation.out:1239]: super-nit: It would be nice if these errors did not come from another node, but instead the privileges were checked before sending commands there. Now users would get this "confusing" `CONTEXT:  while executing command on localhost:xxxxx`

Probably not important enough to do in this PR. So let's make an issue, seems like a "good first issue".

File: src/test/regress/spec/isolation_database_cmd_from_any_node.spec [hunk 28]
```diff
@@ -0,0 +1,89 @@
+session "s1"
+
+step "s1-begin" { BEGIN; }
+step "s1-commit" { COMMIT; }
+step "s1-rollback" { ROLLBACK; }
+step "s1-enable-create-db-prop" { SET citus.enable_create_database_propagation TO ON; }
```
Comment (JelteF) [inline src/test/regress/spec/isolation_database_cmd_from_any_node.spec:6]: let's do this in per-session `setup` for both sessions. i.e. like this:
```
session "s1"
setup { SET citus.enable_create_database_propagation TO ON; }

step "s1-begin" ....
```

File: src/test/regress/expected/isolation_database_cmd_from_any_node.out [hunk 29]
```diff
@@ -0,0 +1,235 @@
...
+step s2-acquire-citus-adv-oclass-lock-with-oid-testdb2: SELECT citus_internal_acquire_citus_advisory_object_class_lock((SELECT CASE WHEN substring(version(), '\d+')::integer < 16 THEN 25 ELSE 26 END AS oclass_database), 'testdb2');
+citus_internal_acquire_citus_advisory_object_class_lock
+---------------------------------------------------------------------
+
+(1 row)
+
+step s1-commit: COMMIT;
+step s2-commit: COMMIT;
+step s1-drop-testdb1: DROP DATABASE testdb1;
+step s2-drop-testdb2: DROP DATABASE testdb2;
```
Comment (JelteF) [inline src/test/regress/expected/isolation_database_cmd_from_any_node.out:66]: a shared `teardown` with `DROP DB IF EXISTS` for all the databases would be nice. Right now the cleanup distracts a bit from the actual test

File: src/test/regress/spec/isolation_database_cmd_from_any_node.spec [hunk 30]
```diff
@@ -0,0 +1,89 @@
...
+
+step "s1-create-user-dbuser" { CREATE USER dbuser; }
+step "s1-drop-user-dbuser" { DROP USER dbuser; }
+
+step "s1-acquire-citus-adv-oclass-lock" { SELECT citus_internal_acquire_citus_advisory_object_class_lock((SELECT CASE WHEN substring(version(), '\d+')::integer < 16 THEN 25 ELSE 26 END AS oclass_database), NULL); }
+step "s1-acquire-citus-adv-oclass-lock-with-oid-testdb1" { SELECT citus_internal_acquire_citus_advisory_object_class_lock((SELECT CASE WHEN substring(version(), '\d+')::integer < 16 THEN 25 ELSE 26 END AS oclass_database), 'testdb1'); }
```
Comment (JelteF) [inline src/test/regress/spec/isolation_database_cmd_from_any_node.spec:12]: Maybe add a comment that the object class changed from 25 to 26 in PG16

**Review by JelteF — APPROVED** [review]
Apart from removing (see other comment for details):
```
case OCLASS_SUBSCRIPTION:
```
And a few other very minor things I think this is good to merge.

File: src/test/regress/expected/isolation_database_cmd_from_any_node.out (hunk 29 above)
Comment (onurctirtir) [reply src/test/regress/expected/isolation_database_cmd_from_any_node.out:66]: Such a teardown block would mean an implicit transaction block and DROP DATABASE cannot be executed in a transaction block, so unfortunately cannot have a common teardown block :/

teardown
{
    DROP DATABASE IF EXISTS testdb1;
    DROP DATABASE IF EXISTS testdb2;
    DROP DATABASE IF EXISTS db1;
}

Comment (JelteF) [reply src/test/regress/expected/isolation_database_cmd_from_any_node.out:66]: ugh, I forgot about that. Let's leave as is then.

File: src/test/regress/expected/create_drop_database_propagation.out (hunk 27 above)
Comment (onurctirtir) [reply src/test/regress/expected/create_drop_database_propagation.out:1239]: let me quickly fix that in this PR while we're at it
